### ISSUE TITLE
python binding: dump() writes to stdout, instead of returning value

### ISSUE THREAD

**Issue by azrdev** [oetiker/rrdtool-1.x#676]
Try:

    import rrdtool
    xml = rrdtool.dump('some-db.rrd')
Expected: `xml` should contain XML file/string
Actual: XML is dumped to stdout, `xml` is empty

**Comment by oetiker** [comment]
rrdtool dump actually DOES write to stdout ... in order to get the dump in memory the code of rrd_dump.c would have to be altered first!

**Comment by nirgal** [comment]
Maybe the python documentation could be improved:
```
$ pydoc rrdtool.dump

Help on built-in function dump in rrdtool:
   rrdtool.dump = dump(...)
    dump - dump an RRD to XML
    [--header|-h {none,xsd,dtd}] [--no-header]file.rrd [file.xml]
```

**Comment by azrdev** [comment]
@oetiker yes, and since this is not really useful for a library it's obviously a bug

**Comment by oetiker** [comment]
Are you suggesting that the dump functionality should not have been included into the python bindings. As you are expecting 'dump' to not dump into a file but into memory ? 

If you are working on a patch, please make it so that it preserves the present functionality since many people my be relying on this. Just adding a memory dump function (since rrd files can be pretty large, you may want to be careful with this as it may cause undue pressure on the memory subsystem. The restore functionality for example has been re-written to use a pull parser to avoid reading in the complete rrd file as this was causing problems on some platforms)

**Comment by atommaki** [comment]
I agree with @azrdev, the current state is useless and confusing. (and surprisingly no change in 5+ years...). No one except a function to print on the stdout instead of returning something. If you think the memory usage can be an issue then the dump should have 2 parameters, one for the input (what is has already) and one for the output xml file.

**Comment by oetiker** [comment]
PR will be gladly accepted!

**Comment by atommaki** [comment]
While checking the code I've realized it's actually works, the documentation is not so detailed...

First I've tried like this:
```
rrdtool.dump("test.rrd")
```
the result goes to the stdout, quite useless.

But this one also works and writes the output to a file:
```
rrdtool.dump("test.rrd", "rrdtest.xml")
```
So I would say the bug is invalid, but the documentation also could be more detailed.

**Comment by oetiker** [comment]
### Triage check (master @ 17d1f842)

**Status:** Works as designed / not a bug

The underlying `rrd_dump()` C function always writes to stdout (or an output file) — there is no return-string API at this layer. The Python binding at `_rrdtool_dump()` in `bindings/python/rrdtoolmodule.c` correctly calls `rrd_dump()` and returns `None` on success; capturing the output as a Python string would require using the `rrd_dump_cb_r()` callback API and wiring up a buffer, which is a new feature rather than a bug fix. The existing behaviour of writing to stdout or to a named file (e.g. `rrdtool.dump("test.rrd", "out.xml")`) is consistent with the underlying C library design.

A PR that adds an optional in-memory return path via `rrd_dump_cb_r()` would be welcome, but the current behaviour is by design.
